arwen/elrondapi/elrondei.go

```diff
@@ -3222,24 +3236,70 @@ func v1_4_getReturnDataSize(context unsafe.Pointer, resultID int32) int32 {
 
 //export v1_4_getReturnData
 func v1_4_getReturnData(context unsafe.Pointer, resultID int32, dataOffset int32) int32 {
+	host := arwen.GetVMHost(context)
+
+	result := GetReturnDataWithHostAndTypedArgs(host, resultID)
+	if result == nil {
+		return 0
+	}
+
 	runtime := arwen.GetRuntimeContext(context)
-	output := arwen.GetOutputContext(context)
-	metering := arwen.GetMeteringContext(context)
+	err := runtime.MemStore(dataOffset, result)
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return 0
+	}
+
+	return int32(len(result))
+}
+
+func GetReturnDataWithHostAndTypedArgs(host arwen.VMHost, resultID int32) []byte {
+	output := host.Output()
+	metering := host.Metering()
 
 	gasToUse := metering.GasSchedule().ElrondAPICost.GetReturnData
 	metering.UseGasAndAddTracedGas(getReturnDataName, gasToUse)
 
 	returnData := output.ReturnData()
 	if resultID >= int32(len(returnData)) {
-		return 0
+		return nil
 	}
 
-	err := runtime.MemStore(dataOffset, returnData[resultID])
-	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
-		return 0
-	}
+	return returnData[resultID]
```

Comment:
`resultID` can be negative here. There should be a check that `resultID` is ≥ 0, or at least cast it to an `uint32`. Contracts written in C can pass -4 here, causing a panic.